From 4474c7b683fefbfa249200dcfb17b9dbebe5f892 Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Fri, 28 Feb 2014 17:15:30 +0800
Subject: [PATCH 2/2] powerpc/mm: adjust the workaround of Erratum 4801 to
 only for silicon Rev1.0

For the Freescale Erratum 004801 "software thread synchronization is required
for MMU register accesses", which has been fixed on silicon Rev2.0. So adjust
the workaround only for silicon Rev1.0.

Signed-off-by: Yang Wei <Wei.Yang@windriver.com>
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 arch/powerpc/mm/tlb_nohash_low.S |   12 ++++++++++++
 1 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/mm/tlb_nohash_low.S b/arch/powerpc/mm/tlb_nohash_low.S
index 71b67ee..1ef2c69 100644
--- a/arch/powerpc/mm/tlb_nohash_low.S
+++ b/arch/powerpc/mm/tlb_nohash_low.S
@@ -322,6 +322,12 @@ ALT_MMU_FTR_SECTION_END_IFCLR(MMU_FTR_USE_TLBILX)
 
 #ifdef CONFIG_FSL_ERRATUM_A_004801
 .macro tlb_lock
+	mfspr   r0, SPRN_PVR
+	rlwinm r0, r0, 0, 24, 15
+	lis r9, PVR_VER_E6500
+	ori r9, r9, PVR_REV_E6500V1
+	cmpw r0, r9
+	bne 1f
 	ld	r7,PACA_TLB_PER_CORE_PTR(r13)
 	mtocrf	0x01,r7
 	addi	r8,r7,PACA_TLB_LOCK-1 /* -1 to compensate for low bit set */
@@ -342,6 +348,12 @@ ALT_MMU_FTR_SECTION_END_IFCLR(MMU_FTR_USE_TLBILX)
 .endm
 
 .macro tlb_unlock
+	mfspr   r0, SPRN_PVR
+	rlwinm r0, r0, 0, 24, 15
+	lis r9, PVR_VER_E6500
+	ori r9, r9, PVR_REV_E6500V1
+	cmpw r0, r9
+	bne 1f
 	mtocrf	0x01,r7
 	bf	31,1f		/* no lock if TLB_PER_CORE_HAS_LOCK clear */
 	li	r9,0
-- 
1.7.5.4

